File: sdk/go/opensecstack/webhook_test.go
package opensecstack

import (
	"context"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
	"time"
)

// signBody returns the "sha256=<hex>" HMAC-SHA256 signature for body using secret.
func signBody(body []byte, secret string) string {
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write(body)
	return "sha256=" + hex.EncodeToString(mac.Sum(nil))
}

// ---------------------------------------------------------------------------
// 1. VerifySignature — valid signature
// ---------------------------------------------------------------------------

func TestVerifySignature_Valid(t *testing.T) {
	secret := "supersecret"
	body := []byte(`{"id":"evt_001","event_type":"apiguard.scan.completed"}`)
	sig := signBody(body, secret)

	if err := VerifySignature(body, sig, secret); err != nil {
		t.Fatalf("expected nil error for valid signature, got: %v", err)
	}
}

// ---------------------------------------------------------------------------
// 2. VerifySignature — wrong secret
// ---------------------------------------------------------------------------

func TestVerifySignature_WrongSecret(t *testing.T) {
	body := []byte(`{"id":"evt_002","event_type":"apiguard.scan.failed"}`)
	sig := signBody(body, "correct-secret")

	err := VerifySignature(body, sig, "wrong-secret")
	if err == nil {
		t.Fatal("expected ErrInvalidSignature, got nil")
	}
	if err != ErrInvalidSignature {
		t.Fatalf("expected ErrInvalidSignature, got: %v", err)
	}
}

// ---------------------------------------------------------------------------
// 3. VerifySignature — tampered body
// ---------------------------------------------------------------------------

func TestVerifySignature_TamperedBody(t *testing.T) {
	secret := "mysecret"
	original := []byte(`{"id":"evt_003","event_type":"citadel.hard_stop"}`)
	sig := signBody(original, secret)

	tampered := []byte(`{"id":"evt_003","event_type":"citadel.vigil_red"}`)
	err := VerifySignature(tampered, sig, secret)
	if err == nil {
		t.Fatal("expected ErrInvalidSignature for tampered body, got nil")
	}
	if err != ErrInvalidSignature {
		t.Fatalf("expected ErrInvalidSignature, got: %v", err)
	}
}

// ---------------------------------------------------------------------------
// 4. WebhookRouter — dispatches to the correct handler
// ---------------------------------------------------------------------------

func TestWebhookRouter_DispatchesToCorrectHandler(t *testing.T) {
	secret := "routersecret"
	var got WebhookEvent

	router := NewWebhookRouter(secret)
	router.On(EventAPIScanCompleted, func(_ context.Context, e WebhookEvent) error {
		got = e
		return nil
	})
	router.On(EventCITADELHardStop, func(_ context.Context, e WebhookEvent) error {
		t.Error("wrong handler called: hard_stop handler should not fire")
		return nil
	})

	event := WebhookEvent{
		ID:        "evt_004",
		EventType: EventAPIScanCompleted,
		Source:    "apiguard",
		Timestamp: time.Now().UTC(),
		Payload:   json.RawMessage(`{"scan_id":"scan-abc"}`),
	}
	body, _ := json.Marshal(event)
	sig := signBody(body, secret)

	req := httptest.NewRequest(http.MethodPost, "/webhooks", strings.NewReader(string(body)))
	req.Header.Set("X-APIGuard-Signature", sig)
	rec := httptest.NewRecorder()

	router.ServeHTTP(rec, req)

	if rec.Code != http.StatusOK {
		t.Fatalf("expected 200, got %d: %s", rec.Code, rec.Body.String())
	}
	if got.ID != "evt_004" {
		t.Fatalf("handler received wrong event ID: got %q, want %q", got.ID, "evt_004")
	}
	if got.EventType != EventAPIScanCompleted {
		t.Fatalf("handler received wrong event type: %q", got.EventType)
	}
}

// ---------------------------------------------------------------------------
// 5. WebhookRouter — returns 400 on bad signature
// ---------------------------------------------------------------------------

func TestWebhookRouter_BadSignatureReturns400(t *testing.T) {

	router := NewWebhookRouter("the-real-secret")
	router.On(EventCITADELVigilRed, func(_ context.Context, e WebhookEvent) error {
		t.Error("handler should not be called with a bad signature")
		return nil
	})

	body := []byte(`{"id":"evt_005","event_type":"citadel.vigil_red","source":"citadel"}`)
	badSig := signBody(body, "not-the-real-secret")

	req := httptest.NewRequest(http.MethodPost, "/webhooks", strings.NewReader(string(body)))
	req.Header.Set("X-Citadel-Signature", badSig)
	rec := httptest.NewRecorder()

	router.ServeHTTP(rec, req)

	if rec.Code != http.StatusBadRequest {
		t.Fatalf("expected 400 for bad signature, got %d", rec.Code)
	}
}

// ---------------------------------------------------------------------------
// 6. WebhookRouter — catch-all handler fires for unregistered event types
// ---------------------------------------------------------------------------

func TestWebhookRouter_CatchAll(t *testing.T) {
	secret := "catchsecret"
	var catchedType string

	router := NewWebhookRouter(secret)
	router.On("*", func(_ context.Context, e WebhookEvent) error {
		catchedType = e.EventType
		return nil
	})

	event := WebhookEvent{
		ID:        "evt_catch",
		EventType: "some.unknown.event",
		Source:    "apiguard",
		Timestamp: time.Now().UTC(),
		Payload:   json.RawMessage(`{}`),
	}
	body, _ := json.Marshal(event)
	sig := signBody(body, secret)

	req := httptest.NewRequest(http.MethodPost, "/webhooks", strings.NewReader(string(body)))
	req.Header.Set("X-APIGuard-Signature", sig)
	rec := httptest.NewRecorder()

	router.ServeHTTP(rec, req)

	if rec.Code != http.StatusOK {
		t.Fatalf("expected 200, got %d", rec.Code)
	}
	if catchedType != "some.unknown.event" {
		t.Errorf("catch-all received wrong event type: %q", catchedType)
	}
}

// ---------------------------------------------------------------------------
// 7. WebhookRouter — no handler registered returns 200 silently
// ---------------------------------------------------------------------------

func TestWebhookRouter_NoHandler_Returns200(t *testing.T) {
	secret := "nosecret"
	router := NewWebhookRouter(secret) // no handlers registered

	event := WebhookEvent{
		ID:        "evt_nohandler",
		EventType: EventNIS2ControlUpdated,
		Source:    "nis2compass",
		Timestamp: time.Now().UTC(),
		Payload:   json.RawMessage(`{}`),
	}
	body, _ := json.Marshal(event)
	sig := signBody(body, secret)

	req := httptest.NewRequest(http.MethodPost, "/webhooks", strings.NewReader(string(body)))
	req.Header.Set("X-Citadel-Signature", sig)
	rec := httptest.NewRecorder()

	router.ServeHTTP(rec, req)

	if rec.Code != http.StatusOK {
		t.Fatalf("expected 200 when no handler registered, got %d", rec.Code)
	}
}

// ---------------------------------------------------------------------------
// 8. WebhookRouter — handler error returns 500
// ---------------------------------------------------------------------------

func TestWebhookRouter_HandlerError_Returns500(t *testing.T) {
	secret := "errsecret"
	router := NewWebhookRouter(secret)
	router.On(EventCITADELVigilAmber, func(_ context.Context, _ WebhookEvent) error {
		return fmt.Errorf("downstream service unavailable")
	})

	event := WebhookEvent{
		ID:        "evt_err",
		EventType: EventCITADELVigilAmber,
		Source:    "citadel",
		Timestamp: time.Now().UTC(),
		Payload:   json.RawMessage(`{}`),
	}
	body, _ := json.Marshal(event)
	sig := signBody(body, secret)

	req := httptest.NewRequest(http.MethodPost, "/webhooks", strings.NewReader(string(body)))
	req.Header.Set("X-Citadel-Signature", sig)
	rec := httptest.NewRecorder()

	router.ServeHTTP(rec, req)

	if rec.Code != http.StatusInternalServerError {
		t.Fatalf("expected 500 on handler error, got %d", rec.Code)
	}
}

File: sdk/typescript/src/__tests__/webhook.test.ts
import { describe, it, expect, vi } from "vitest";
import { createHmac } from "node:crypto";
import {
  WebhookRouter,
  verifySignature,
  InvalidSignatureError,
  EventAPIScanCompleted,
  EventAPIScanFailed,
  EventCITADELHardStop,
} from "../webhook.js";
import type { WebhookEvent } from "../webhook.js";

// ─── Helpers ────────────────────────────────────────────────────────────────

function sign(body: Buffer | string, secret: string): string {
  const mac = createHmac("sha256", secret);
  mac.update(typeof body === "string" ? Buffer.from(body) : body);
  return "sha256=" + mac.digest("hex");
}

function makeEventJSON(overrides?: Partial<WebhookEvent>): string {
  const event: WebhookEvent = {
    id: "evt-001",
    event_type: EventAPIScanCompleted,
    source: "apiguard",
    timestamp: "2026-03-31T12:00:00Z",
    payload: { scan_id: "scan-001", findings_count: 3 },
    ...overrides,
  };
  return JSON.stringify(event);
}

const SECRET = "test-webhook-secret";

// ─── verifySignature ────────────────────────────────────────────────────────

describe("verifySignature", () => {
  it("1. accepts a valid signature", () => {
    const body = Buffer.from("hello world");
    const sig = sign(body, SECRET);
    expect(() => verifySignature(body, sig, SECRET)).not.toThrow();
  });

  it("2. rejects an invalid signature", () => {
    const body = Buffer.from("hello world");
    expect(() =>
      verifySignature(body, "sha256=0000000000000000000000000000000000000000000000000000000000000000", SECRET),
    ).toThrow(InvalidSignatureError);
  });

  it("3. rejects a tampered body", () => {
    const body = Buffer.from("original");
    const sig = sign(body, SECRET);
    const tampered = Buffer.from("tampered");
    expect(() => verifySignature(tampered, sig, SECRET)).toThrow(
      InvalidSignatureError,
    );
  });

  it("4. rejects wrong secret", () => {
    const body = Buffer.from("test");
    const sig = sign(body, "correct-secret");
    expect(() => verifySignature(body, sig, "wrong-secret")).toThrow(
      InvalidSignatureError,
    );
  });

  it("5. rejects empty signature", () => {
    const body = Buffer.from("test");
    expect(() => verifySignature(body, "", SECRET)).toThrow(
      InvalidSignatureError,
    );
  });
});

// ─── WebhookRouter ──────────────────────────────────────────────────────────

describe("WebhookRouter", () => {
  it("6. dispatches to the correct handler", async () => {
    const handler = vi.fn();
    const router = new WebhookRouter(SECRET);
    router.on(EventAPIScanCompleted, handler);

    const body = Buffer.from(makeEventJSON());
    const sig = sign(body, SECRET);

    const event = await router.handle(body, sig);

    expect(handler).toHaveBeenCalledOnce();
    expect(handler).toHaveBeenCalledWith(
      expect.objectContaining({ event_type: EventAPIScanCompleted }),
    );
    expect(event.id).toBe("evt-001");
  });

  it("7. rejects bad signature with InvalidSignatureError", async () => {
    const router = new WebhookRouter(SECRET);
    router.on(EventAPIScanCompleted, vi.fn());

    const body = Buffer.from(makeEventJSON());

    await expect(router.handle(body, "sha256=bad")).rejects.toThrow(
      InvalidSignatureError,
    );
  });

  it("8. catch-all handler fires for unregistered event types", async () => {
    const catchAll = vi.fn();
    const router = new WebhookRouter(SECRET);
    router.on("*", catchAll);

    const body = Buffer.from(
      makeEventJSON({ event_type: "custom.unknown.event" }),
    );
    const sig = sign(body, SECRET);

    await router.handle(body, sig);

    expect(catchAll).toHaveBeenCalledOnce();
    expect(catchAll).toHaveBeenCalledWith(
      expect.objectContaining({ event_type: "custom.unknown.event" }),
    );
  });

  it("9. no handler registered — returns event without error", async () => {
    const router = new WebhookRouter(SECRET);
    const body = Buffer.from(makeEventJSON());
    const sig = sign(body, SECRET);

    const event = await router.handle(body, sig);
    expect(event.event_type).toBe(EventAPIScanCompleted);
  });

  it("10. handler error propagates", async () => {
    const router = new WebhookRouter(SECRET);
    router.on(EventAPIScanCompleted, () => {
      throw new Error("handler failed");
    });

    const body = Buffer.from(makeEventJSON());
    const sig = sign(body, SECRET);

    await expect(router.handle(body, sig)).rejects.toThrow("handler failed");
  });

  it("11. method chaining", () => {
    const router = new WebhookRouter(SECRET);
    const result = router
      .on(EventAPIScanCompleted, vi.fn())
      .on(EventAPIScanFailed, vi.fn())
      .on(EventCITADELHardStop, vi.fn());
    expect(result).toBe(router);
  });

  it("12. async handler is awaited", async () => {
    let resolved = false;
    const router = new WebhookRouter(SECRET);
    router.on(EventAPIScanCompleted, async () => {
      await new Promise((r) => setTimeout(r, 10));
      resolved = true;
    });

    const body = Buffer.from(makeEventJSON());
    const sig = sign(body, SECRET);

    await router.handle(body, sig);
    expect(resolved).toBe(true);
  });

  it("13. specific handler takes precedence over catch-all", async () => {
    const specific = vi.fn();
    const catchAll = vi.fn();
    const router = new WebhookRouter(SECRET);
    router.on(EventAPIScanCompleted, specific).on("*", catchAll);

    const body = Buffer.from(makeEventJSON());
    const sig = sign(body, SECRET);

    await router.handle(body, sig);

    expect(specific).toHaveBeenCalledOnce();
    expect(catchAll).not.toHaveBeenCalled();
  });

  it("14. event constants match Go/Python values", () => {
    expect(EventAPIScanCompleted).toBe("apiguard.scan.completed");
    expect(EventAPIScanFailed).toBe("apiguard.scan.failed");
    expect(EventCITADELHardStop).toBe("citadel.hard_stop");
  });
});
